File: android/src/main/java/com/nabidreams/lib/audio/RecorderModule.kt
package com.nabidreams.lib.audio

import com.facebook.react.bridge.*
import com.facebook.react.modules.core.DeviceEventManagerModule
import java.io.IOException

class RecorderModule(private val reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {
    enum class EventType(val value: String) {
        STATE_CHANGE("recorderStateChange")
    }

    override fun getName(): String {
        return "Recorder"
    }

    override fun getConstants(): MutableMap<String, Any> {
        return mapOf(
                "State" to Recorder.State.values().map { it.name to it.value }.toMap(),
                "EventType" to EventType.values().map { it.name to it.value }.toMap(),
                "MIN_LEVEL" to Recorder.MIN_POWER,
                "MAX_LEVEL" to Recorder.MAX_POWER
        ).toMutableMap()
    }

    @ReactMethod
    fun getLevel(promise: Promise) {
        promise.resolve(recorder.maxPower)
    }

    @ReactMethod
    fun getState(promise: Promise) {
        promise.resolve(recorder.state.value)
    }

    @ReactMethod
    fun start(filePath: String, promise: Promise) {
        try {
            recorder.start(filePath)
            promise.resolve(null)
        } catch (e: IOException) {
            promise.reject(e)
        }
    }

    @ReactMethod
    fun stop(promise: Promise) {
        recorder.stop()
        promise.resolve(null)
    }

    private val recorder: Recorder = Recorder().apply {
        stateChangeListener = { state ->
            sendEvent(EventType.STATE_CHANGE.value, Arguments.makeNativeMap(mapOf("state" to state.value)))
        }
    }

    private fun sendEvent(eventName: String, data: Any? = null) {
        reactContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java).emit(eventName, data)
    }
}

File: ios/RecorderModule.swift
@objc(Recorder)
class RecorderModule: RCTEventEmitter {
  enum EventType: String {
    case stateChange = "recorderStateChange"
  }
  
  override static func requiresMainQueueSetup() -> Bool {
    return false
  }
  
  override func constantsToExport() -> [AnyHashable: Any]? {
    return [
      "State": [
        "STARTED": Recorder.State.started.rawValue,
        "STOPPED": Recorder.State.stopped.rawValue,
      ],
      
      "EventType": [
        "STATE_CHANGE": EventType.stateChange.rawValue
      ],
      
      "MIN_LEVEL": Recorder.minPower,
      "MAX_LEVEL": Recorder.maxPower,
    ]
  }
  
  override func supportedEvents() -> [String]! {
    return [
      EventType.stateChange.rawValue
    ]
  }
  
  override init() {
    super.init()

    recorder.stateChangeListener = { state in
      self.sendEvent(withName: EventType.stateChange.rawValue, body: ["state": state.rawValue])
    }
  }
  
  @objc
  func getLevel(_ resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
    resolve(recorder.averagePower)
  }
  
  @objc
  func getState(_ resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
    resolve(recorder.state.rawValue)
  }
  
  @objc
  func start(_ filePath: String, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
    do {
      try recorder.start(filePath)
      resolve(nil)
    } catch {
      reject("Error", error.localizedDescription, error)
    }
  }
  
  @objc
  func stop(_ resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
    do {
      try recorder.stop()
      resolve(nil)
    } catch {
      reject("Error", error.localizedDescription, error)
    }
  }
  
  private let recorder: Recorder! = Recorder()
}
